### PR TITLE
[CARBONDATA-1827] S3 Carbon Implementation

### PR DESCRIPTION
**Pull request by jatin9896**
1) Provide support for s3 in carbondata.
2) Added S3Example to create store on s3.
3) Added S3CSVExample to load csv from s3.
 

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed? NO
 
 - [x] Any backward compatibility impacted? NO
 
 - [x] Document update required? NO

 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required? 
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       Added Examples to test the functionality
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1566/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2801/

**Comment by anubhav100** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2896/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1571/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2807/

**Comment by anubhav100** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2899/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1572/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2808/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2810/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1574/

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    CarbonProperties.getInstance()
+      .addProperty(CarbonCommonConstants.CARBON_TIMESTAMP_FORMAT, "yyyy/MM/dd HH:mm:ss")
+      .addProperty(CarbonCommonConstants.CARBON_DATE_FORMAT, "yyyy/MM/dd")
+      .addProperty(CarbonCommonConstants.ENABLE_UNSAFE_COLUMN_PAGE_LOADING, "true")
+      .addProperty(CarbonCommonConstants.DEFAULT_CARBON_MAJOR_COMPACTION_SIZE, "0.02")
+
+    import org.apache.spark.sql.CarbonSession._
+    if (args.length != 4) {
+      logger.error("Usage: java CarbonS3Example <access-key> <secret-key>" +
+                   "<carbon store location> <spark-master>")
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: arg(2) is table path, not store path, right?

Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: you are right. ok I will change it to tablepath.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+object S3Example {
+
+  /**
+   * This example demonstrate usage of s3 as a store.
+   *
+   * @param args require three parameters "Access-key" "Secret-key"
+   *             "s3 bucket path" "spark-master"
+   */
+
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: remove empty line

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+import org.apache.spark.sql.SparkSession
+import org.slf4j.{Logger, LoggerFactory}
+
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.apache.carbondata.core.util.CarbonProperties
+
+object S3Example {
+
+  /**
+   * This example demonstrate usage of s3 as a store.
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: Change to 
> This example demonstrate usage of
> 1. create carbon table with storage location on object based storage like AWS S3, Huawei OBS, etc
> 2. load data into carbon table, the generated file will be stored on object based storage
> 3. query the table.
> With the indexing feature of carbondata, the data read from object based storage is minimum, thus providing both high performance analytic and low cost storage

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  def main(args: Array[String]) {
+    val rootPath = new File(this.getClass.getResource("/").getPath
+                            + "../../../..").getCanonicalPath
+    val path = s"$rootPath/examples/spark2/src/main/resources/data1.csv"
+    val logger: Logger = LoggerFactory.getLogger(this.getClass)
+    CarbonProperties.getInstance()
+      .addProperty(CarbonCommonConstants.CARBON_TIMESTAMP_FORMAT, "yyyy/MM/dd HH:mm:ss")
+      .addProperty(CarbonCommonConstants.CARBON_DATE_FORMAT, "yyyy/MM/dd")
+      .addProperty(CarbonCommonConstants.ENABLE_UNSAFE_COLUMN_PAGE_LOADING, "true")
+      .addProperty(CarbonCommonConstants.DEFAULT_CARBON_MAJOR_COMPACTION_SIZE, "0.02")
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: remove unnecssary property setting, all these are not required

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+       """.stripMargin)
+
+    spark.sql(
+      s"""
+         | LOAD DATA LOCAL INPATH '$path'
+         | INTO TABLE carbon_table
+         | OPTIONS('HEADER'='true')
+       """.stripMargin)
+
+    spark.sql("ALTER table carbon_table compact 'MINOR'")
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: Add a comment: 
> // Use compaction command to merge segments or small files in object based storage, this can be done periodically.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+  def main(args: Array[String]) {
+    val rootPath = new File(this.getClass.getResource("/").getPath
+                            + "../../../..").getCanonicalPath
+    val logger: Logger = LoggerFactory.getLogger(this.getClass)
+
+    CarbonProperties.getInstance()
+      .addProperty(CarbonCommonConstants.CARBON_TIMESTAMP_FORMAT, "yyyy/MM/dd HH:mm:ss")
+      .addProperty(CarbonCommonConstants.CARBON_DATE_FORMAT, "yyyy/MM/dd")
+      .addProperty(CarbonCommonConstants.ENABLE_UNSAFE_COLUMN_PAGE_LOADING, "true")
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala]: remove unnecssary property setting, all these are not required

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    spark.sql("show segments for table carbon_table").show()
+
+    spark.sql(
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: remove this to make it simpler

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 8]
```diff
@@ -168,6 +168,14 @@
 
   public static final String S3A_PREFIX = "s3a://";
 
+  public static final String S3N_ACCESS_KEY = "fs.s3n.awsAccessKeyId";
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: Add comment for these 4 properties

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+object S3CsvExample {
+
+  /**
+   * This example demonstrate to create local store and load data from CSV files on S3
+   *
+   * @param args require three parameters "Access-key" "Secret-key"
+   *             "s3 path to csv" "spark-master"
+   */
+
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala]: remove empty line

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2906/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1585/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2820/

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala (hunk 6 above)
Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala]: Removed all properties.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 5 above)
Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: Added

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 4 above)
Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: done

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 3 above)
Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: done

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 2 above)
Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: removed.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala (hunk 9 above)
Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala]: done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1598/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2916/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1615/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2851/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1621/

**Comment by anubhav100** [comment]
retest this please

**Comment by jackylk** [comment]
Can you show an example of how to give <table_path> in S3Example using AWS S3? I tried with Huawei OBS, it has two problems:
1. I need to set the endpoint conf in main function by `spark.sparkContext.hadoopConfiguration.set("fs.s3a.endpoint", "obs.cn-north-1.myhwclouds.com")
` manually
2. After I set the conf, there is an exception thrown when running it: 
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/http/pool/ConnPoolControl

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2936/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2863/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1630/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2882/

**Comment by jatin9896** [comment]
@jackylk table path is the path to bucket location like I have provided s3a://<bucket-name>/<location> and regarding endpoints, I have modified example which takes endpoint as args(4) and it is not mandatory to provide. About connection pooling exception in the example is also fixed. Please check.

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1650/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2892/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1661/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2895/

**Comment by anubhav100** [comment]
retest this please

**Comment by jackylk** [comment]
I tried, it is successful now. But in log it is showing using LOCAL_SORT, it should be NO_SORT, right?

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+      .builder()
+      .master(args(3))
+      .appName("S3Example")
+      .config("spark.driver.host", "localhost")
+      .config(accessKey, args(0))
+      .config(secretKey, args(1))
+      .config(endpoint, getS3EndPoint(args))
+      .getOrCreateCarbonSession()
+
+    spark.sparkContext.setLogLevel("INFO")
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: change to WARN, it is printing too many

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+       """.stripMargin)
+
+    spark.sql(
+      s"""
+         | LOAD DATA LOCAL INPATH '$path'
+         | INTO TABLE carbon_table
+         | OPTIONS('HEADER'='true')
+       """.stripMargin)
+
+    spark.sql(
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: do a select * after 1 load

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+      s"""
+         | LOAD DATA LOCAL INPATH '$path'
+         | INTO TABLE carbon_table
+         | OPTIONS('HEADER'='true')
+       """.stripMargin)
+
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: Now we do not allow doing compaction and data load concurrently, it will raise exception: Cannot run data loading and compaction on same table concurrently. Please wait for load to finish

So, I think you should check the finish of data load then do compaction, you can check finish of data load by SHOW SEGMENT and collect the result, loop until there are 3 segments

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    // Use compaction command to merge segments or small files in object based storage,
+    // this can be done periodically.
+    spark.sql("ALTER table carbon_table compact 'MINOR'")
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: change this to MAJOR, and remove subsequent command, to make it simpler

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+       """.stripMargin)
+
+    spark.sql(
+      s"""
+         | LOAD DATA LOCAL INPATH '$path'
+         | INTO TABLE carbon_table
+         | OPTIONS('HEADER'='true')
+       """.stripMargin)
+
+    spark.sql(
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: remove one load to make it simpler

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+
+object S3Example {
+
+  /**
+   * This example demonstrate usage of
+   * 1. create carbon table with storage location on object based storage
+   *    like AWS S3, Huawei OBS, etc
+   * 2. load data into carbon table, the generated file will be stored on object based storage
+   *    query the table.
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: remove this line

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+   * 3. With the indexing feature of carbondata, the data read from object based storage is minimum,
+   *    thus providing both high performance analytic and low cost storage
+   *
+   * @param args require three parameters "Access-key" "Secret-key"
+   *             "s3 bucket path" "spark-master" "s3-endpoint"
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: modify to  `table path on S3`

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+  def main(args: Array[String]) {
+    val rootPath = new File(this.getClass.getResource("/").getPath
+                            + "../../../..").getCanonicalPath
+    val path = s"$rootPath/examples/spark2/src/main/resources/data1.csv"
+    val logger: Logger = LoggerFactory.getLogger(this.getClass)
+
+    import org.apache.spark.sql.CarbonSession._
+    if (args.length < 4 || args.length > 5) {
+      logger.error("Usage: java CarbonS3Example <access-key> <secret-key>" +
+                   "<table-path> <spark-master> <s3-endpoint>")
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: Modify to `<table-path-on-S3>`

Comment (jackylk) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: move <spark-master> as the last parameter, it should be optional, default value is 'local'

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1679/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2912/

**Comment by jackylk** [comment]
please rebase and drop the first two commits which are merged to carbonstore branch already

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2961/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1703/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2937/

**Comment by jackylk** [comment]
LGTM

**Comment by jatin9896** [comment]
retest this please

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/AlterTableLoadPartitionRDD.scala [hunk 18]
```diff
@@ -41,102 +41,95 @@ class AlterTableLoadPartitionRDD[K, V](alterPartitionModel: AlterPartitionModel,
...
 
-    var storeLocation: String = null
-    val carbonLoadModel = alterPartitionModel.carbonLoadModel
-    val segmentId = alterPartitionModel.segmentId
-    val oldPartitionIds = alterPartitionModel.oldPartitionIds
-    val carbonTable = carbonLoadModel.getCarbonDataLoadSchema.getCarbonTable
-    val databaseName = carbonTable.getDatabaseName
-    val factTableName = carbonTable.getTableName
-    val partitionInfo = carbonTable.getPartitionInfo(factTableName)
+  var storeLocation: String = null
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/AlterTableLoadPartitionRDD.scala]: Can you put these codestyle modification into a separate PR?

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+
+    while (countSegment.length != 3) {
+      this.wait(2000)
+    }
+
+    // Use compaction command to merge segments or small files in object based storage,
+    // this can be done periodically.
+    spark.sql("ALTER table carbon_table compact 'MAJOR'")
+    spark.sql("show segments for table carbon_table").show()
+
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: please do one more select * query

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+    } else if (path.startsWith(CarbonCommonConstants.S3_PREFIX)) {
+      ("spark.hadoop." + CarbonCommonConstants.S3_ACCESS_KEY,
+        "spark.hadoop." + CarbonCommonConstants.S3_SECRET_KEY, endPoint)
+    } else {
+      throw new Exception("Incorrect Store Path")
+    }
+  }
+
+  def getS3EndPoint(args: Array[String]): String = {
+    if (args.length >= 4 && args(3).contains(".com")) {
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: why need to compare with ".com"?

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+      args(3)
+    }
+    else {
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: move one line up

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+      ""
+    }
+  }
+
+  def getSparkMaster(args: Array[String]): String = {
+    if (args.length >= 4) {
+      if (args.length == 5) {
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: this logic can be optimized, you can make one level of `if... else if`

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+        args(4)
+      }
+      else if (args(3).contains("spark:") || args(3).contains("mesos:")) {
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: `else if` should be same line as `}`

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+        args(3)
+      }
+      else {
+        "local"
+      }
+    }
+    else {
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: `else if` should be same line as `}`

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 20 above)
Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: it might happen that at args(4) be spark master.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 22 above)
Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: sure

Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: done

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 21 above)
Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: done

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 19 above)
Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: added one more select query.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 23 above)
Comment (jatin9896) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: updated to same line.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1714/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2948/

**Comment by jackylk** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2953/

**Comment by jackylk** [comment]
It is merged, please close it manually

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2978/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1724/
